codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -925,5 +920,306 @@ final void incrementStreamableBytesForTree(int numBytes) {
+                if (isWritableConnection(connectionState)) {
+                    // If the write operation does not occur we still need to check all streams because they
+                    // may have transitioned from writable to not writable.
+                    checkAllWritabilityChanged(connectionState);
+                }
+            }
+        }
+
+        @Override
+        public void enqueueFrame(AbstractState state, FlowControlled frame) throws Http2Exception {
+            super.enqueueFrame(state, frame);
+            checkConnectionThenStreamWritabilityChanged(connectionState(), state);
+        }
+
+        @Override
+        public boolean isWritable(AbstractState state) {
+            return isWritableConnection(connectionState()) && state.windowSize() - state.pendingBytes() > 0;
+        }
+
+        @Override
+        public Listener listener() {
+            return listener;
+        }
+
+        @Override
+        public void stateCancelled(AbstractState state) {
+            try {
+                checkConnectionThenStreamWritabilityChanged(connectionState(), state);
+            } catch (Http2Exception e) {
+                logger.error("Caught unexpected exception from checkAllWritabilityChanged", e);
+            }
+        }
+
+        @Override
+        public void channelWritabilityChange() throws Http2Exception {
+            AbstractState connectionState = connectionState();
+            if (connectionState.markWritabilty() != isChannelWritable()) {
+                checkAllWritabilityChanged(connectionState);
+            }
+        }
+
+        private void notifyWritabilityChanged(AbstractState state) {
+            state.markWritabilty(!state.markWritabilty());
+            try {
+                listener.writabilityChanged(state.stream);
+            } catch (RuntimeException e) {
+                logger.error("Caught unexpected exception from listener.writabilityChanged", e);
+            }
+        }
+
+        private void checkConnectionThenStreamWritabilityChanged(AbstractState connectionState, AbstractState state)
+                throws Http2Exception {
+            // It is possible that the connection window and/or the individual stream writability could change.
+            if (isWritableConnection(connectionState) != connectionState.markWritabilty()) {
+                checkAllWritabilityChanged(connectionState);
+            } else if (isWritable(state) != state.markWritabilty()) {
+                notifyWritabilityChanged(state);
+            }
+        }
+
+        private void checkAllWritabilityChanged(AbstractState connectionState) throws Http2Exception {
+            // Make sure we mark that we have notified as a result of this change.
+            connectionState.markWritabilty(isWritableConnection(connectionState));
+            connection.forEachActiveStream(checkStreamWritabilityVisitor);
+        }
+
+        private boolean isWritableConnection(AbstractState connectionState) {
```

Comment:
maybe just make a connectionState member, rather than passing it around everywhere?